### PR TITLE
Add bufstyle linter

### PR DESCRIPTION
**Pull request by amckinney**
This adds the `bufstyle` linter to enforce Buf's Go code standards.

For now, this sets up the framework for adding new `go/analysis.Analyzer` implementations to the `bufstyle` linter, and starts with a simple `packagefilename` check that verifies that every package has a file with the same name (e.g. package `foo` has `foo.go`).

Note that the output isn't nearly as fancy as `golangci-lint` - it's just raw text. For example:
```sh
$ make lint
...
go install ./private/pkg/licenseheader/cmd/license-header
go install ./private/pkg/git/cmd/git-ls-files-unstaged
license-header --license-type apache --copyright-holder Buf Technologies, Inc. --year-range 2020-2022 ALL_FILES
golangci-lint run --timeout 3m0s
/path/to/bufstyle ./private/buf/... ./private/bufpkg/... ./private/bufstyle/... ./private/pkg/... 2>&1
~/bufteam/buf/private/buf/cmd/buf/workspacetests/other/proto/workspacetest/doc.go:18:1: Package "workspacetest" does not have a workspacetest.go
make[2]: *** [bufstylelint] Error 3
make[1]: *** [shortlint] Error 2
make: *** [lint] Error 2
```

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -4,6 +4,7 @@
 .env/
 .tmp/
 .vscode/
+bin/
```
[the remark this answers is not in the document]
Comment (amckinney) [reply .dockerignore:7]: Yea I figured this one would come up. Happy to come up with alternatives here - we can discuss.

**Review by bufdev — COMMENTED** [review]
I'm sure you thought about this but I really don't like this plugin build mode - this really should be a separate binary that we can `go install` if at all possible

**Comment by amckinney** [comment]
> I'm sure you thought about this but I really don't like this plugin build mode - this really should be a separate binary that we can go install if at all possible

Yea, that'll actually make it portable too - it's only runnable with `golangci-lint` as-is. I'll see if I can build this as a separate binary that we can run independent of `golangci-lint`.

File: make/go/dep_bufstyle.mk [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+$(call _assert_var,CACHE_BIN)
+
+# Unlike 'buf', we always install bufstyle from source since
+# we don't release it as an independent binary.
+.PHONY: $(BUFSTYLE)
+$(BUFSTYLE):
+	GOBIN=$(CACHE_BIN) go install ./cmd/bufstyle
+
+# Use this instead of "buf" when using buf.
+BUFSTYLE_BIN := $(CACHE_BIN)/bufstyle
```
Comment (amckinney) [inline make/go/dep_bufstyle.mk:15]: @bufdev I'll need your eyes on this.

**Review by bufdev — APPROVED** [review]
Approving my own refactors, although I'm not that cool

**Pull request merged**